### ISSUE TITLE
Training with Tacotron GTA mel specs

### ISSUE THREAD

**Issue by duvtedudug** [r9y9/wavenet_vocoder#52]
I've generated a set of ground-truth aligned mel spectrograms from Rahane's Tacotron-2.

I've trained for over 100k steps but still getting poor results for longer sequences...
[gta-taco-wave-120k.zip](https://github.com/r9y9/wavenet_vocoder/files/1938729/gta-taco-wave-120k.zip)

Any ideas on how to improve this?

**Comment by Rayhane-mamah** [comment]
I think you're in the same case as:

https://github.com/Rayhane-mamah/Tacotron-2/issues/29

**Comment by duvtedudug** [comment]
@Rayhane-mamah thanks for the quick reply.

But I think my case is different, since I am using aligned spectrograms output from your Tacotron (not real spectrograms generated from training audio)

??

**Comment by Rayhane-mamah** [comment]
Oh in that case it's something worth looking at! :)

Could you report tacotron model params? Reduction factor, training steps, etc?

If it's caused by the feature prediction network I'll find the issue.

**Comment by duvtedudug** [comment]
I think the spectrograms from your Tacotron are good. I trained for 137000 steps on the standard settings (no changes in hparams) and synthesised GTA. 

Here's the plots...

![step-137000-real-mel-spectrogram](https://user-images.githubusercontent.com/298918/39141156-ccaccd28-471e-11e8-82d1-7c6be2597666.png)
![step-137000-align](https://user-images.githubusercontent.com/298918/39141157-ccccf814-471e-11e8-8b7c-2bf94e860731.png)
![step-137000-pred-mel-spectrogram](https://user-images.githubusercontent.com/298918/39141159-cce672a8-471e-11e8-856b-78a0f7859dc1.png)

When training wavenet vocoder on real spectrograms I get decent results after about 50k steps (8bit softmax)

But in examples I posted previously I'm not getting good quality even after 100k+ steps

??

**Comment by duvtedudug** [comment]
Ah of course. I should have noticed that. Thanks for the pretrained model I will try with r=1

**Comment by r9y9** [comment]
A little off topic: I'm wondering if you get such a smooth and natural (at a glance) mel-spectrogram in fact. Did you use https://github.com/Rayhane-mamah/Tacotron-2/blob/1547b2502305f4ee58bceede1384054c22b0497a/tacotron/utils/plot.py#L36-L38 for plotting mel-spectrogram? If so, can you try additional params `interpolation="none"` to `imshow`?

```
plt.imshow(np.rot90(spectrogram))
```

```
plt.imshow(np.rot90(spectrogram), interpolation="none")
```

I remember I had hard time to get smooth mel-spectrogram when I was working on DeepVoice3 (even on Tacotron). I think RNNs can do better than CNNs but I'm curious if Tacotron2 actually performs better than Deepvoice3 and Tacotron.

**Comment by duvtedudug** [comment]
@r9y9 @Rayhane-mamah correct me if I'm wrong. If I use external mel spectrograms from Taco2, then I only need to make sure the Taco2 specs have the same shape as the wavenet vocoder specs generated from preprocess? (i.e. same sample rate, same num hops, 80 dimensional) I presume they do not need to have the exact same frequency ranges or amplitudes?

**Comment by duvtedudug** [comment]
Here's quick check with no interpolation:
<img width="1381" alt="specs_check_interp_none" src="https://user-images.githubusercontent.com/298918/39180299-64a12c16-47ae-11e8-9723-ecb5b88fa04c.png">

**Comment by r9y9** [comment]
@duvtedudug Thanks! Looks good. I should definitely check out the details of Tacotron2.

**Comment by duvtedudug** [comment]
@r9y9 @Rayhane-mamah no problem. Can anybody confirm my question above? re: Taco specs shape only requirement?

**Comment by duvtedudug** [comment]
Thanks!

@r9y9 is there a wavenet limitation for local conditioning to be in 0, 1 ?

Maybe I should be more patient and train for 200-300k steps?

**Comment by Rayhane-mamah** [comment]
Yeah further training might also be the cause, since mels are noisier than
real ones and are in a wider range ( I'm supposing weights are initialized
close to 0 ).

I will leave the final words to the wavenet expert however :)

**Comment by r9y9** [comment]
There's no limitation for the range of conditional features. Notice that current implementation assumes the range of [0, 1] for simplicity; https://github.com/r9y9/wavenet_vocoder/blob/4d5f68c8341eb1dad1307ada7d99c8059840ca64/wavenet_vocoder/wavenet.py#L162-L164.

**Comment by r9y9** [comment]
> I've generated a set of ground-truth aligned mel spectrograms from Rahane's Tacotron-2.
> I've trained for over 100k steps but still getting poor results for longer sequences...
> Any ideas on how to improve this?

Are you sure you did correct time resolution adjustment as I did in https://github.com/r9y9/deepvoice3_pytorch/blob/3226e415ef1d8412bb159b228aa3c9212fdb892e/generate_aligned_predictions.py#L38-L42? Also did you use exactly same audio feature extraction pipeline with Tacotron2 and WaveNet? If both you did correctly, then I think you should just be more patient. As I mentioned in https://github.com/r9y9/wavenet_vocoder/issues/45#issuecomment-383313651, it will generally take 1000k steps to get sufficient good quality with MoL output layer.

**Comment by duvtedudug** [comment]
> Are you sure you did correct time resolution adjustment 

The main difference I could see is the WaveNet specs have extra padding of a few frames at the end. I have padded the Taco generated specs with silence, to match the audio files of WaveNet preprocess. Otherwise audio is identical and mel timing is aligned. I have ```assert taco.shape==wavenet.shape``` before training.

There is difference in the LWS stft of WaveNet seems to be in range [0.4, 0.8] approx ?

I have approximated this [0.4, 0.8] with scaling.

> it will generally take 1000k steps to get sufficient good quality with MoL 

I'm using 256 mulaw quantised for quicker results (on real specs starts to sound good 50k-100k for me)

**Comment by duvtedudug** [comment]
[train_eval.zip](https://github.com/r9y9/wavenet_vocoder/files/1949842/train_eval.zip)

train_eval audio showing some promise. 190k file not there yet but 180k file sounds not too bad.

**Comment by butterl** [comment]
@duvtedudug  could you help to share your hpara settings in wavenet to train the synthesised GTA npy from Tacotron 2 ?

**Comment by duvtedudug** [comment]
@butterl Normal hparam settings just changed to 256, mulaw-quantised

**Comment by PetrochukM** [comment]
@r9y9 Why did you include RELu? The Wavenet paper does not discuss using RELu in the transposition network.

https://github.com/r9y9/wavenet_vocoder/blob/740219b5ce8d7a96c30fb68189727451abf9539d/wavenet_vocoder/wavenet.py#L161-L163
